### PR TITLE
New exchange: vertex

### LINKED ISSUE #18682: New Exchange: Vertex (DEX)
### Exchange Name

Vertex

### URL of the Exchange

https://vertexprotocol.com/

### URL of the Exchange's API docs

https://vertex-protocol.gitbook.io/docs/getting-started/overview

### Why should this exchange be implemented?

Because DEX

### PR DESCRIPTION
**Pull request by sc0Vu**
New decentralized exchange vertex protocol (https://app.vertexprotocol.com/).

TODO:
- rest

  - public
    - ...

  - private
    - ...

- ws

  - public
    - ...

  - private
    - ...

- test

closes #18682

### CODE REVIEW

File: ts/src/vertex.ts [hunk 1]
```diff
@@ -0,0 +1,1741 @@
...
+            const triggerPrice = this.safeStringN (trigger, [ 'price_above', 'price_below', 'last_price_above', 'last_price_below' ]);
+            if (triggerPrice !== undefined) {
+                triggerPriceNum = this.parseToNumeric (this.convertFromX18 (triggerPrice));
+            }
+        }
+        market = this.safeMarket (marketId, market);
+        const symbol = market['symbol'];
+        let priceNum = undefined;
+        if (price !== undefined) {
+            priceNum = this.parseToNumeric (this.convertFromX18 (price));
```
Comment (carlosmiei) [inline ts/src/vertex.ts:1317]: we don't need to convert amount/price/etc to numeric, they can remain as strings and safeOrder will parse them accordingly.

File: ts/src/vertex.ts [hunk 2]
```diff
@@ -0,0 +1,1741 @@
...
+         * @method
+         * @name vertex#fetchOrders
+         * @description fetches information on multiple orders made by the user
+         * @see https://docs.vertexprotocol.com/developer-resources/api/gateway/queries/orders
+         * @see https://docs.vertexprotocol.com/developer-resources/api/trigger/queries/list-trigger-orders
+         * @param {string} symbol unified market symbol of the market orders were made in
+         * @param {int} [since] the earliest time in ms to fetch orders for
+         * @param {int} [limit] the maximum number of order structures to retrieve
+         * @param {object} [params] extra parameters specific to the exchange API endpoint
+         * @param {boolean} [params.stop] whether the order is a stop/algo order
```
Comment (carlosmiei) [inline ts/src/vertex.ts:1409]: we should use params.trigger, .stop is deprecated

File: ts/src/vertex.ts [hunk 3]
```diff
@@ -0,0 +1,1741 @@
...
+            request['place_order']['trigger'] = trigger;
+            response = await this.v1TriggerPostExecute (this.extend (request, params));
+        } else {
+            response = await this.v1GatewayPostExecute (this.extend (request, params));
+        }
+        //
+        // {
+        //     "status": "success",
+        //     "signature": {signature},
+        //     "data": {
```
Comment (carlosmiei) [inline ts/src/vertex.ts:1231]: there are a lot of trailing whitespaces 😅 
image

File: ts/src/vertex.ts [hunk 4]
```diff
@@ -0,0 +1,2796 @@
...
+        let price = undefined;
+        let amount = undefined;
+        let side = undefined;
+        let fee = undefined;
+        const id = this.safeString2 (trade, 'trade_id', 'submission_idx');
+        const order = this.safeString (trade, 'digest');
+        const timestamp = this.safeTimestamp (trade, 'timestamp');
+        if (timestamp === undefined) {
+            // fetchMyTrades
+            const baseBalance = this.safeDict (this.safeDict (trade, 'pre_balance', {}), 'base', {});
```
Comment (carlosmiei) [inline ts/src/vertex.ts:658]: @sc0Vu fetchMyTrades is not returning amount/price/cost/etc
image

File: ts/src/vertex.ts [hunk 5]
```diff
@@ -0,0 +1,2796 @@
...
+        //     "updated_at": 1688768157050
+        // }
+        //
+        let marketId = this.safeString (order, 'product_id');
+        let timestamp = this.safeTimestamp (order, 'placed_at');
+        let amount = this.safeString (order, 'amount');
+        let price = this.safeString (order, 'price_X18');
+        const remaining = this.safeString (order, 'unfilled_amount');
+        let triggerPriceNum = undefined;
+        const status = this.safeValue (order, 'status');
```
Comment (carlosmiei) [inline ts/src/vertex.ts:1764]: why safeValue here?

Comment (sc0Vu) [reply ts/src/vertex.ts:1765]: @carlosmiei the status would be string and object.

Comment (carlosmiei) [reply ts/src/vertex.ts:1765]: ok ok

**Comment by carlosmiei** [comment]
vertex-logo

**Pull request merged**